### PR TITLE
[MPS] Add native multi-dimensional reduction support for torch.all and torch.any

### LINKED ISSUE #176826: [MPS] Add support for multi-dimensional all / any reductions
The MPS backend does not currently provide a dedicated native implementation for multi-dimensional reductions in `torch.all` and `torch.any` when `dim` is provided as multiple dimensions.

Example:

```python
x = torch.ones(2, 3, 4, device="mps", dtype=torch.bool)

torch.all(x, dim=[0, 1])
torch.any(x, dim=[1, 2])
```

These operations currently work through the composite implementation, which performs reductions sequentially across dimensions. While each individual reduction may use an existing MPS kernel, there is no native MPS implementation that reduces across multiple dimensions in a single operation.

Adding a native implementation for this case would improve backend feature parity with CPU and CUDA and could allow the reduction to be performed more efficiently on MPS.

I have a working implementation with tests and would be happy to open a PR for this.

cc @jerryzh168 @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk

### PR DESCRIPTION
**Pull request by AneeshGidda**
Fixes #176826

## Summary
Adds native MPS support for multi-dimensional reductions in `torch.all` and `torch.any` when `dim` is provided as multiple dimensions.

Previously, this case relied on the composite fallback implementation, which performs sequential reductions across dimensions. This change performs the reduction across multiple axes directly using a single MPSGraph reduction operation.

While the existing implementation may internally use MPS kernels for each reduction step, there is no dedicated implementation for reducing across multiple dimensions in a single operation. This PR adds native support for that case.

The primary motivation of this change is to improve backend feature parity between MPS and other backends.

## Implementation
The implementation introduces native MPS kernels for:

- `all_dims_out_mps`
- `any_dims_out_mps`

These kernels perform reductions across multiple axes using the corresponding MPSGraph operations.

Edge cases handled:

- `dim=None`
- `dim=[]`
- empty tensors
- `keepdim=True` and `keepdim=False`

The implementation falls back to the default behavior for unsupported cases.

## Benchmark
Initial benchmarking was performed comparing this implementation against the previous composite reduction path.

In practice, performance differences between the two approaches were minimal, with results generally falling within normal run-to-run variance. In some cases the new implementation showed a slight improvement, but overall performance is largely comparable.

One potential advantage of the new implementation is reduced intermediate tensor creation, since the reduction can be expressed as a single operation rather than a sequence of reductions. In theory this could reduce memory overhead.

Given these observations, the primary benefit of this change is backend feature parity and a more direct mapping of multi-dimensional reductions to the underlying MPSGraph operations.

## Tests
Added comprehensive test coverage for multi-dimensional `all` and `any` operations:

- `test_all_any_multi_dims`: Tests various combinations of multi-dimensional reductions across different tensor shapes (2D, 3D, 4D) with different dimension combinations
- `test_all_any_dims_none_and_empty`: Tests edge cases with `dim=None` and empty dimension lists
- `test_all_any_negative_dims`: Tests negative dimension indexing for multi-dimensional reductions
- `test_all_any_duplicate_dims_error`: Tests error handling for duplicate dimensions

All tests verify correctness by comparing MPS results against CPU reference implementations.

## Notes
This change improves backend feature parity for the MPS backend while providing a native implementation for multi-dimensional logical reductions.

### CODE REVIEW

**Review by malfet — CHANGES_REQUESTED** [review]
And where is the benchmark code?

File: test/test_mps.py [hunk 1]
```diff
@@ -13401,6 +13385,836 @@ def test_metal_compiler_bug_workaround(self):
         for i in [0, 5, 6, 7, 63, 64]:
             self.assertEqual(out[i], 0)
 
+class TestAllAnyDimsMPS(TestCaseMPS):
```
Comment (malfet) [inline test/test_mps.py]: Why all of that is needed?

File: aten/src/ATen/native/mps/operations/ReduceOps.mm [hunk 2]
```diff
@@ -1430,6 +1432,164 @@ static void all_any_common_impl_mps(const Tensor& input_t,
...
+
+  for (const auto i : c10::irange(num_input_dims)) {
+    const bool is_reduce_dim =
+        std::find(wrapped_dims.begin(), wrapped_dims.end(), static_cast<int64_t>(i)) != wrapped_dims.end();
+    [apparent_out_shape addObject:(is_reduce_dim ? @1 : @(input_shape[i]))];
+  }
+
+  @autoreleasepool {
+    // Build a cache key so identical graphs can be reused
+    std::string dims_str;
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/ReduceOps.mm]: This string can be built cleanly and efficiently in one line in one line using fmt::join from fmtlib.

File: aten/src/ATen/native/mps/operations/ReduceOps.mm [hunk 3]
```diff
@@ -1430,6 +1432,164 @@ static void all_any_common_impl_mps(const Tensor& input_t,
...
+    for (const auto d : wrapped_dims) {
+      dims_str += std::to_string(d) + ",";
+    }
+
+    std::string key = op_name + "_dims_out_mps:" + dims_str + ":" +
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/ReduceOps.mm]: Likewise try fmt::fmt here, way cleaner and more efficient.

File: test/test_mps.py (hunk 1 above)
Comment (AneeshGidda) [reply test/test_mps.py]: I have reduced the tests to the necessary cases and refactored them. I will push the update shortly.

File: aten/src/ATen/native/mps/operations/ReduceOps.mm (hunk 2 above)
Comment (AneeshGidda) [reply aten/src/ATen/native/mps/operations/ReduceOps.mm]: Thanks for the suggestion, will switch to fmt::join

File: aten/src/ATen/native/mps/operations/ReduceOps.mm (hunk 3 above)
Comment (AneeshGidda) [reply aten/src/ATen/native/mps/operations/ReduceOps.mm]: Thanks, will update this to use fmt::format

**Comment by AneeshGidda** [comment]
closing in favour of #180752, which replaces this with a native metal kernel implementation
